src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ParameterConverter.cs

```diff
@@ -44,43 +59,93 @@ internal static class ParameterConverter
 
             if (!literal.Contains("\""))
             {
-                if (string.Equals(literal, "true", StringComparison.OrdinalIgnoreCase))
+                if (TryResolveBooleanValue(literal, out bool parsedBool))
                 {
-                    return true;
+                    return parsedBool;
                 }
-
-                if (string.Equals(literal, "false", StringComparison.OrdinalIgnoreCase))
-                {
-                    return false;
-                }
-
-                if (string.Equals(literal, "null", StringComparison.OrdinalIgnoreCase))
+                if (TryResolveNullValue(literal, out _))
                 {
                     return null;
                 }
-
                 if ((literal.Contains(CultureInfo.CurrentCulture.NumberFormat.NumberDecimalSeparator)
-                     || literal.Contains(CultureInfo.InvariantCulture.NumberFormat.NumberDecimalSeparator))
-                    && ParserExtensions.DoubleTryParseСurrentOrInvariant(literal, out double literalDouble))
+                    || literal.Contains(CultureInfo.InvariantCulture.NumberFormat.NumberDecimalSeparator))
+                    && TryResolveFloatValue(literal, out double parsedFloat))
                 {
-                    return literalDouble;
+                    return parsedFloat;
                 }
-
-                if (long.TryParse(literal, out long literalLong))
+                if (TryResolveIntegerValue(literal, out long parsedInteger))
                 {
-                    return literalLong;
+                    return parsedInteger;
                 }
-
-                if (literal.StartsWith("0x", StringComparison.OrdinalIgnoreCase)
-                    && long.TryParse(literal.Substring(2), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out literalLong))
+                if (TryResolveHexValue(literal, out long parsedHex))
                 {
-                    return literalLong;
+                    return parsedHex;
                 }
             }
-
             return literal;
         }
 
+        /// <summary>
+        /// Tries to convert <paramref name="literal"/> to <paramref name="dataType"/>.
+        /// If <paramref name="dataType"/> is null or empty, the type to be inferred. See <see cref="InferTypeAndConvertLiteral(string)"/> for more details.
+        /// </summary>
+        internal static bool TryConvertLiteralToDatatype(string literal, string? dataType, out object? value)
```

Comment:
I'd say it's time to get rid of `OnParameterError` and then this function can be complately reused within `DataTypeSpecifiedConvertLiteral`. (if we'd want to maintain back compat of that already deprecated functionality - we can just simply first call it if `TryConvertLiteralToDatatype` returns false).

However - it doesn't necesarily fall into scope of this PR